pkg/graph/builder.go

```diff
@@ -682,117 +683,87 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
 // we evaluate B's CEL expressions against 2 emulated resources A and C, and so
 // on.
 func validateResourceCELExpressions(resources map[string]*Resource, instance *Resource) error {
-	resourceNames := maps.Keys(resources)
-	// We also want to allow users to refer to the instance spec in their expressions.
-	resourceNames = append(resourceNames, "schema")
+	resourceNames := []string{"schema"}
 	conditionFieldNames := []string{"schema"}
 
-	env, err := krocel.DefaultEnvironment(krocel.WithResourceIDs(resourceNames))
-	if err != nil {
-		return fmt.Errorf("failed to create CEL environment: %w", err)
-	}
-	instanceEmulatedCopy := instance.emulatedObject.DeepCopy()
-	if instanceEmulatedCopy != nil && instanceEmulatedCopy.Object != nil {
-		delete(instanceEmulatedCopy.Object, "apiVersion")
-		delete(instanceEmulatedCopy.Object, "kind")
-		delete(instanceEmulatedCopy.Object, "status")
-	}
-
 	for _, resource := range resources {
-		for _, resourceVariable := range resource.variables {
-			for _, expression := range resourceVariable.Expressions {
-				err := validateCELExpressionContext(env, expression, resourceNames)
-				if err != nil {
-					return fmt.Errorf("failed to validate expression context: '%s' %w", expression, err)
-				}
+		// First validate includeWhen expressions
+		for _, includeWhenExpression := range resource.includeWhenExpressions {
+			instanceEnv, err := krocel.DefaultEnvironment(krocel.WithResourceIDs(resourceNames))
+			if err != nil {
+				return fmt.Errorf("failed to create CEL environment: %w", err)
+			}
```

Comment:
Why are we instantiating a CEL environment for every CEL expression?

Reply:
Creating a new CEL environment for every expression is inefficient. We should implement caching of CEL environments.